instrumentation/action_pack/lib/opentelemetry/instrumentation/action_pack/handlers/action_controller.rb

```diff
@@ -0,0 +1,61 @@
+# frozen_string_literal: true
+
+# Copyright The OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module Instrumentation
+    module ActionPack
+      module Handlers
+        # Action controller handler to handle the notification from ActiveSupport
+        class ActionController
+          # @param config [Hash] of instrumentation options
+          def initialize(config)
+            @config = config
+          end
+
+          # Invoked by ActiveSupport::Notifications at the start of the instrumentation block
+          #
+          # @param _name [String] of the Event (unused)
+          # @param _id [String] of the event (unused)
+          # @param payload [Hash] containing job run information
+          # @return [Hash] the payload passed as a method argument
+          def start(_name, _id, payload)
+            rack_span = OpenTelemetry::Instrumentation::Rack.current_span
+
+            # from rails 6.1, the request will be added to payload
+            request = payload[:request]
+            request = payload[:headers].instance_variable_get(:@req) if ::ActionPack.version < Gem::Version.new('6.1.0')
+
+            rack_span.name = "#{payload[:controller]}##{payload[:action]}" unless request.env['action_dispatch.exception']
+
+            attributes_to_append = {
+              OpenTelemetry::SemanticConventions::Trace::CODE_NAMESPACE => String(payload[:controller]),
+              OpenTelemetry::SemanticConventions::Trace::CODE_FUNCTION => String(payload[:action])
+            }
+
+            attributes_to_append[OpenTelemetry::SemanticConventions::Trace::HTTP_TARGET] = request.filtered_path if request.filtered_path != request.fullpath
+
+            rack_span.add_attributes(attributes_to_append)
+          rescue StandardError => e
+            OpenTelemetry.handle_error(exception: e)
+          end
+
+          # Invoked by ActiveSupport::Notifications at the end of the instrumentation block
+          #
+          # @param _name [String] of the Event (unused)
+          # @param _id [String] of the event (unused)
+          # @param payload [Hash] containing job run information
+          # @return [Hash] the payload passed as a method argument
+          def finish(_name, _id, payload)
+            rack_span = OpenTelemetry::Instrumentation::Rack.current_span
+            rack_span.record_exception(payload[:exception_object]) if payload[:exception_object]
```

Comment:
what about `action_dispatch.exception`? I see you reference it in start but not record it?

Reply:
> what about action_dispatch.exception? I see you reference it in start but not record it?

I had difficulty producing the `request.env['action_dispatch.exception']` from request.env although I set the exceptions_app. But I think `request.env['action_dispatch.exception']` is the same object as `payload[:exception_object]`.

In fact, if rename the span if controller and action are present, then there is no need to check `request.env['action_dispatch.exception']`

Reply:
Sounds good. I think that makes sense to me.

Reply:
Just realized that the `action_dispatch.exception` will only appear if there is no exceptions_app configured. This `action_dispatch.exception` will be added in [show_exception](https://github.com/rails/rails/blob/main/actionpack/lib/action_dispatch/middleware/show_exceptions.rb#L37), however, [debug_exception](https://github.com/rails/rails/blob/main/actionpack/lib/action_dispatch/middleware/debug_exceptions.rb#L43) choose not to raise the exception to show_exceptions if `wrapper.show?(request)` is true.

Anyway, `action_dispatch.exception` and `payload[:exception_object]` should have same object; and if renaming the span with present of controller and action, then there is no need to check `action_dispatch.exception`.

Reply:
`debug_exception` is only for dev. mode right? Does `debug_exception` end up causing span leaks because it does not call `finish` on the notifications?

Reply:
I think `debug_exception` is for all environment based on [default_middleware_stack.rb](https://github.com/rails/rails/blob/main/railties/lib/rails/application/default_middleware_stack.rb#L57-L59). 

> Does debug_exception end up causing span leaks because it does not call finish on the notifications?

I don't think so, since this actionpack instrumentation only modify the existing rack span, so rack instrumentation will make sure span won't be unfinished/leaked. Also, the notification is subscribed on `process_action`, and debug_exception is just part of the entire middleware stack, therefore, I think the notification will finish.

Reply:
🤦‍♂️ of course. I keep building this mental model where this instrumentation makes its own spans